### ISSUE TITLE
preemptive resources with schedule to reduce capacity

### ISSUE THREAD

**Issue by tomjemmett** [r-simmer/simmer#246]
I'm having some issues with preemptive resources and schedules that reduce the capacity of the resource over time. I've simplified my issue to the reprex below; while my actual problem uses multiple generators with different priorities it's clearer to see the issue with a single entity, all with the same priority.  The two examples below are `e1` (with `preemptive = FALSE`) and `e2` (with `preemptive = TRUE`).

`e1` behaves as I expect - the first 3 entities are able to start straight away, the 4th entity has to wait. At time = 4 we reduce the resource to 1, but because `y0`-`y2` have already started they continue until completion. `y3` starts once `y2` completes, and `y4` doesn't get a chance to start because at time = 10 we reduce the resource to 0. `y3` continues running even after time = 10 until it is complete.

`e2` behaves oddly - when we reduce the resource at time = 4 `y0` and `y1` immediately stop running (even though they arrived before `y2`. Once `y2` completes, `y0` restarts and completes. It's hard to tell from this, but I think `y1` restarts at time = 8, but is interupted again at time = 10 and doesn't complete (as there are no resources), or perhaps one of `y3`/`y4` starts, but does not have enough time to complete before the capacity is reduce to 0 and the activity is interupted.

I'm not sure if this is intended behaviour, or a bug? If it is intended it would be useful to have an argument to allow you to continue already started activities until completion.

``` r
library(simmer)
trj <- trajectory() %>%
  seize("x") %>%
  timeout(5) %>%
  release("x")

e1 <- simmer() %>%
  add_resource("x", schedule(c(0, 4, 10), c(3, 1, 0))) %>%
  add_generator("y", trj, from_to(0, 5, function() 1)) %>%
  run(16)

e1 %>% get_mon_arrivals(FALSE, TRUE)
#>   name start_time end_time activity_time finished replication
#> 1   y0          0        5             5     TRUE           1
#> 2   y1          1        6             5     TRUE           1
#> 3   y2          2        7             5     TRUE           1
#> 4   y3          3       12             5     TRUE           1
#> 5   y4          4       NA            NA    FALSE           1

e2 <- simmer() %>%
  add_resource("x", schedule(c(0, 4, 10), c(3, 1, 0)), preemptive = TRUE) %>%
  add_generator("y", trj, from_to(0, 5, function() 1)) %>%
  run(16)

e2 %>% get_mon_arrivals(FALSE, TRUE)
#>   name start_time end_time activity_time finished replication
#> 1   y2          2        7             5     TRUE           1
#> 2   y0          0        8             5     TRUE           1
#> 3   y4          4       NA            NA    FALSE           1
#> 4   y3          3       NA            NA    FALSE           1
#> 5   y1          1       NA            NA    FALSE           1
```

**Comment by Enchufa2** [comment]
If you want see a full trace of what's happening, set `verbose=TRUE` in your `simmer()` call.

TL;DR, yes, it's intended behaviour. Enabling preemption means that, if you shrink your resource, arrivals are preempted. At `t=4`, y0, y1, y2 are running. y0 and y1 are preempted because, by default, preemption occurs in a FIFO basis for equal priorities (see `preempt_order` in `help(add_resources)`). So, when y2 finishes, y0 is processed, and then y1 is restarted, but then capacity is reduced again to 0, so it's preempted again.

So maybe you expected a mixed preemption mode, in which arrivals may preempt other arrivals, but capacity changes do not?

**Comment by tomjemmett** [comment]
yes - looking for mixed preemption where the arrivals (with different priorities) would preempt other arrivals, but when the capacity changes activity is not interupted (similar to the first example). This is the behaviour a colleague had in a model built in simul8 (which we translated to simmer, but were getting different results due to this behaviour) - in the example we have there is a doctor's office where a receptionist is answering phone calls / dealing with emails. The phone calls will preempt the email tasks, but there are a number of shift changes through the day. What we are seeing is at shift changes the receptionists "abandon" the current work (e.g. the call), but the behaviour we expect is that the receptionist would finish the job that they are currently doing before finishing their shift.

**Comment by Enchufa2** [comment]
Suppose we have `capacity=2`, preemption is enabled, and two tasks are being executed. Then `capacity=1`, and the two tasks continue. Then a task with a higher priority arrives. What should happen? (There are more intrincate situations, but let's start with a simple one)

**Comment by tomjemmett** [comment]
If the real world scenario that we are modelling is 2 servers (a and b) and a is "going home" at time point T, but will finish whatever task they have started, then it would be fine for b to be preempted by the higher priority, but not a. (Of course, that leads us to the next question of now what happens if a second higher priority comes in?). It would be great to be able to define some kind of strategy for these scenarios - for instance, the server that is "going home" cannot be preempted, the server that is "staying" can. I suspect that this may have to be some kind of function provided by the user (we don't name the N resources, so which ones are "going home"?)

or, would the recommended way be to split the resources up? something like the below? (I'm not sure if this is the best way of achieving this, but it seems to work. Of note, `value` in `set_capacity_selected()` help doesn't indicate you can pass in a function, but it seems to work! I could send a PR for updated documentation if helpful)

``` r
library(simmer)
e <- simmer(verbose = TRUE)
t <- trajectory() %>%
  select(function() if (now(e) < 4) c("a", "b") else "a") %>%
  seize_selected() %>%
  timeout(5) %>% 
  release_selected() %>%
  set_capacity_selected(function() {
    s <- get_selected(e)
    if ((s == "b" & now(e) >= 4) | (s == "c" & now(e) >= 10)) -1 else 0
  }, mod = "+")

e %>%
  add_resource("a", 2, preemptive = TRUE) %>%
  add_resource("b", 1, preemptive = TRUE) %>%
  add_generator("y", t, from_to(0, 5, function() 1)) %>%
  run(16)
#>          0 |    source: y                |       new: y0               | 0
#>          0 |   arrival: y0               |  activity: Select           | function(), shortest-queue
#>          0 |   arrival: y0               |  activity: Seize            | [0], 1, 0 paths
#>          0 |  resource: a                |   arrival: y0               | SERVE
#>          0 |    source: y                |       new: y1               | 1
#>          0 |   arrival: y0               |  activity: Timeout          | 5
#>          1 |   arrival: y1               |  activity: Select           | function(), shortest-queue
#>          1 |   arrival: y1               |  activity: Seize            | [0], 1, 0 paths
#>          1 |  resource: a                |   arrival: y1               | SERVE
#>          1 |    source: y                |       new: y2               | 2
#>          1 |   arrival: y1               |  activity: Timeout          | 5
#>          2 |   arrival: y2               |  activity: Select           | function(), shortest-queue
#>          2 |   arrival: y2               |  activity: Seize            | [0], 1, 0 paths
#>          2 |  resource: b                |   arrival: y2               | SERVE
#>          2 |    source: y                |       new: y3               | 3
#>          2 |   arrival: y2               |  activity: Timeout          | 5
#>          3 |   arrival: y3               |  activity: Select           | function(), shortest-queue
#>          3 |   arrival: y3               |  activity: Seize            | [0], 1, 0 paths
#>          3 |  resource: a                |   arrival: y3               | ENQUEUE
#>          3 |    source: y                |       new: y4               | 4
#>          4 |   arrival: y4               |  activity: Select           | function(), shortest-queue
#>          4 |   arrival: y4               |  activity: Seize            | [0], 1, 0 paths
#>          4 |  resource: a                |   arrival: y4               | ENQUEUE
#>          5 |   arrival: y0               |  activity: Release          | [0], 1
#>          5 |  resource: a                |   arrival: y0               | DEPART
#>          5 |      task: Post-Release     |          :                  | 
#>          5 |  resource: a                |   arrival: y3               | SERVE
#>          5 |   arrival: y0               |  activity: SetCapacity      | [0], function(), +
#>          5 |   arrival: y3               |  activity: Timeout          | 5
#>          6 |   arrival: y1               |  activity: Release          | [0], 1
#>          6 |  resource: a                |   arrival: y1               | DEPART
#>          6 |      task: Post-Release     |          :                  | 
#>          6 |  resource: a                |   arrival: y4               | SERVE
#>          6 |   arrival: y1               |  activity: SetCapacity      | [0], function(), +
#>          6 |   arrival: y4               |  activity: Timeout          | 5
#>          7 |   arrival: y2               |  activity: Release          | [0], 1
#>          7 |  resource: b                |   arrival: y2               | DEPART
#>          7 |      task: Post-Release     |          :                  | 
#>          7 |   arrival: y2               |  activity: SetCapacity      | [0], function(), +
#>         10 |   arrival: y3               |  activity: Release          | [0], 1
#>         10 |  resource: a                |   arrival: y3               | DEPART
#>         10 |      task: Post-Release     |          :                  | 
#>         10 |   arrival: y3               |  activity: SetCapacity      | [0], function(), +
#>         11 |   arrival: y4               |  activity: Release          | [0], 1
#>         11 |  resource: a                |   arrival: y4               | DEPART
#>         11 |      task: Post-Release     |          :                  | 
#>         11 |   arrival: y4               |  activity: SetCapacity      | [0], function(), +
#> simmer environment: anonymous | now: 11 | next: 
#> { Monitor: in memory }
#> { Resource: a | monitored: TRUE | server status: 0(2) | queue status: 0(Inf) }
#> { Resource: b | monitored: TRUE | server status: 0(0) | queue status: 0(Inf) }
#> { Source: y | monitored: 1 | n_generated: 5 }

get_mon_arrivals(e, TRUE, TRUE)
#>   name start_time end_time activity_time resource replication
#> 1   y0          0        5             5        a           1
#> 2   y1          1        6             5        a           1
#> 3   y2          2        7             5        b           1
#> 4   y3          3       10             5        a           1
#> 5   y4          4       11             5        a           1
get_mon_resources(e) %>%
  tibble::as_tibble()
#> # A tibble: 11 x 9
#>    resource  time server queue capacity queue_size system limit replication
#>    <chr>    <dbl>  <int> <int>    <dbl>      <dbl>  <int> <dbl>       <int>
#>  1 a            0      1     0        2        Inf      1   Inf           1
#>  2 a            1      2     0        2        Inf      2   Inf           1
#>  3 b            2      1     0        1        Inf      1   Inf           1
#>  4 a            3      2     1        2        Inf      3   Inf           1
#>  5 a            4      2     2        2        Inf      4   Inf           1
#>  6 a            5      2     1        2        Inf      3   Inf           1
#>  7 a            6      2     0        2        Inf      2   Inf           1
#>  8 b            7      0     0        1        Inf      0   Inf           1
#>  9 b            7      0     0        0        Inf      0   Inf           1
#> 10 a           10      1     0        2        Inf      1   Inf           1
#> 11 a           11      0     0        2        Inf      0   Inf           1
```

**Comment by Enchufa2** [comment]
> If the real world scenario that we are modelling is 2 servers (a and b) and a is "going home" at time point T, but will finish whatever task they have started, then it would be fine for b to be preempted by the higher priority, but not a. (Of course, that leads us to the next question of now what happens if a second higher priority comes in?). It would be great to be able to define some kind of strategy for these scenarios - for instance, the server that is "going home" cannot be preempted, the server that is "staying" can. I suspect that this may have to be some kind of function provided by the user (we don't name the N resources, so which ones are "going home"?)
> 
> or, would the recommended way be to split the resources up? something like the below? (I'm not sure if this is the best way of achieving this, but it seems to work.

Definitely. As a rule of thumb, whenever two servers can be distinguised in any way (different meanings, different purposes...) the recommendation is to split them in different resources under different names. In your example, it could be `"receptionist_persistent"` and `"receptionist_preemptive"` or whatever, so that in your analysis you are able to just split by "_" and aggregate for "receptionist" if you wish.

> Of note, `value` in `set_capacity_selected()` help doesn't indicate you can pass in a function, but it seems to work! I could send a PR for updated documentation if helpful)

Yes, `value` can be a function. That's an oversight in the docs. A PR would be welcome. :)

**Comment by tomjemmett** [comment]
thanks for your help with this - I'll try to write this up in a blog for my own memory, if it's useful then I could submit as a vignette :-)

**Comment by Enchufa2** [comment]
Please, provide a link here when you have a blog post, and we'll see if it fits as a vignette (or maybe even a [brick](https://github.com/r-simmer/simmer.bricks)?). I'll give a thought to your proposal of allowing preemption from arrivals but not capacity changes, but I'll have to figure out all the corner cases, and I fear this could lead to a rabbit hole.
